### PR TITLE
simple LSTM model implementation

### PR DESCRIPTION
**Pull request by rick020**

### CODE REVIEW

File: sequence/callbacks.py [hunk 1]
```diff
@@ -76,9 +76,14 @@ def callback(**kwargs):
                 packed_padded, padded = ds.get_batch(
                     i * 100, i * 100 + 100, device=device
                 )
+                if model.__class__.__name__ == "LSTM":
```
Comment (ritchie46) [inline sequence/callbacks.py:79]: You could create an args object here. This way you only do the check ` if model.__class__.__name__ == "LSTM"` once.

Something like:
```python
args = [packed_padded]
if model.__class__.__name__ == "LSTM":
    # could also be directly appendedn, however this is more readable
    state_h, state_c = model.init_state(padded.shape[1])
    args.append((state_h, state_c))

with torch.no_grad():
    pred = model(*args)
```

Comment (ritchie46) [inline sequence/callbacks.py:79]: You could create an args object here. This way you only do the check ` if model.__class__.__name__ == "LSTM"` once.

Something like:
```python
args = [packed_padded]
if model.__class__.__name__ == "LSTM":
    # could also be directly appendedn, however this is more readable
    state_h, state_c = model.init_state(padded.shape[1])
    args.append((state_h, state_c))

with torch.no_grad():
    pred = model(*args)
```

File: sequence/model/lstm.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+from torch import nn, optim
+
+import pickle
+import numpy as np
+
+
+class LSTM(Embedding):
+    def __init__(self, vocabulary_size, embedding_dim=10, custom_embeddings=None):
+        self.lstm_layers = 1
```
Comment (ritchie46) [inline sequence/model/lstm.py:14]: maybe add these attributes to the input args? That way you could modify the later.

File: sequence/model/lstm.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+        self.latent_size = 20
+        super().__init__(vocabulary_size, embedding_dim, custom_embeddings)
+
+        self.lstm = nn.LSTM(
+            input_size=embedding_dim,
+            hidden_size=self.latent_size,
+            num_layers=self.lstm_layers,
+            # dropout=0.2,
```
Comment (ritchie46) [inline sequence/model/lstm.py:22]: Uncommented code is considered bad practice. I would also add a dropout parameter to the input of this class so you could modify it.

File: sequence/model/lstm.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+        )
+        self.fc = nn.Linear(self.latent_size, vocabulary_size)
+
+    def forward(self, x, prev_state):
+        emb = self.apply_emb(x)
+        padded_emb, lengths = nn.utils.rnn.pad_packed_sequence(
+            emb, padding_value=0, batch_first=True
+        )
+        output, state = self.lstm(padded_emb, prev_state)
+        fc = self.fc(output)
```
Comment (ritchie46) [inline sequence/model/lstm.py:33]: This is correct. However when you want to add layers to this model it becomes less readable. I would add the `lstm` and the `nn.Linear` layers in a `torch.nn.Sequential` wrapper

File: sequence/model/lstm.py [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+        return torch.log_softmax(fc, -1), state
+
+    def init_state(self, batch):
+        # TODO use random number for init
```
Comment (ritchie46) [inline sequence/model/lstm.py:38]: Why random numbers? During the start of a sequence a nullified hidden state represent that there isn't any hidden state yet.

File: sequence/model/modular.py [hunk 6]
```diff
@@ -20,7 +20,7 @@ def __init__(
             self.vocabulary_size = custom_embeddings.shape[0]
             self.embedding_dim = custom_embeddings.shape[1]
             self.emb = nn.Embedding(*custom_embeddings.shape, _weight=custom_embeddings)  # type: ignore
-            self.emb.weight.requires_grad = False
+            self.emb.weight.requires_grad = True
```
Comment (ritchie46) [inline sequence/model/modular.py:23]: This line is redundant. The default is `True`. But do you want custom embeddings to be trainable?

File: sequence/train/lstm.py [hunk 7]
```diff
@@ -0,0 +1,81 @@
+from sequence.model.lstm import det_loss
+from sequence.utils import backward
+from sequence.callbacks import apply as apply_callbacks
+import logging
+
+logger = logging.getLogger(__name__)
+
+
+def run_epoch(
```
Comment (ritchie46) [inline sequence/train/lstm.py:9]: This code is probably very similar to the existing `run_epoch` functions. Maybe you could generalize and reuse some snippets/ maybe not.

File: sequence/train/stamp.py [hunk 8]
```diff
@@ -53,6 +53,7 @@ def run_epoch(
         epoch_p = i / (n_batches - 1)
         global_step += 1
         c += 1
+
```
Comment (ritchie46) [inline sequence/train/stamp.py:56]: Meaningful change! :wink:

File: sequence/model/lstm.py [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+    model: LSTM,
+    packed_padded: nn.utils.rnn.PackedSequence,
+    test_loss: bool = False,
+    scale_loss_by_lengths: bool = False,
+    max_len: int = 1,
+    state_h: tuple = None,
+    state_c: tuple = None,
+):
+    # b,l,v
+    y_hat, (state_h, state_c) = model(packed_padded, (state_h, state_c))
```
Comment (ritchie46) [inline sequence/model/lstm.py:56]: Seems to me that if you wrap you model in a fuction `Fn(packed_padded) -> y_hat` you could reuse the existing `det_loss` function, right? You could initialize `state_h` and `state_c` in the model implicitly.

**Review by ritchie46 — COMMENTED** [review]
I think some things could be a bit more DRY, but overall seems good. Did it train?

File: sequence/model/lstm.py (hunk 4 above)
Comment (rick020) [reply sequence/model/lstm.py:33]: Like this?

```python
self.wrapper = nn.Sequential([self.lstm, nn.Linear(self.latent_size, vocabulary_size)])
        

    def forward(self, x, prev_state):
        emb = self.apply_emb(x)
        padded_emb, lengths = nn.utils.rnn.pad_packed_sequence(
            emb, padding_value=0, batch_first=True
        )
        output, state = self.wrapper(padded_emb, prev_state)
        return torch.log_softmax(output, -1), state
```

File: sequence/callbacks.py (hunk 1 above)
Comment (rick020) [reply sequence/callbacks.py:79]: The LSTM model return 2 a prediction and a state, so  ```python pred = model(*args)``` results in a tuple. This tuple causes the next lines to file, because it's expecting a prediction

Comment (ritchie46) [reply sequence/callbacks.py:79]: I dont think you need the state. So you could return only y_hat
